### ISSUE TITLE
Did someone delete the mdtraj-1.2.0 binaries from binstar?

### ISSUE THREAD

**Issue by rmcgibbo** [omnia-md/conda-recipes#80]
Many of the mdtraj 1.2.0 binaries have gone missing from the main channel. See in particular the linux and OS-X builds, which now only have py26 versions.
https://binstar.org/omnia/mdtraj/files?version=1.2.0&channel=main

This, for example, has broken the msmbuilder travis. https://travis-ci.org/msmbuilder/msmbuilder/jobs/46279339, which was working as of 15 days ago (the last PR https://travis-ci.org/msmbuilder/msmbuilder/builds/45003144).

**Comment by rmcgibbo** [comment]
Basically, the builds pushed 5 hours ago on the jenkins channel seem to have overwritten the files on the main channel.

**Comment by rmcgibbo** [comment]
The binstar client doesn't appear to allow two files with the same name on different channels.
```
$ binstar upload -u rmcgibbo /Users/rmcgibbo/miniconda/conda-bld/osx-64/mdtraj-1.2.0-np19py27_0.tar.bz2
[...]
Upload(s) Complete
```

```
$ binstar upload -u rmcgibbo -c test-channel /Users/rmcgibbo/miniconda/conda-bld/osx-64/mdtraj-1.2.0-np19py27_0.tar.bz2
[...]
Distribution already exists. Please use the -i/--interactive or --force options or `binstar remove rmcgibbo/mdtraj/1.2.0/osx-64/mdtraj-1.2.0-np19py27_0.tar.bz2`
[Conflict] file osx-64/mdtraj-1.2.0-np19py27_0.tar.bz2 already exists for package mdtraj version 1.2.0
```

Which basically means that the --force being used [on the choderalab jenkins server](https://jenkins.choderalab.org/job/conda-omnia-release-osx-2/lastStableBuild/console) is systematically deleting all of the files I built and pushed to the main channel for our mdtraj release.

**Comment by rmcgibbo** [comment]
The simplest solution might just be to start pushing the builds to the `main` channel. What are the blockers to doing that?

**Comment by jchodera** [comment]
> Basically, the builds pushed 5 hours ago on the jenkins channel seem to have overwritten the files on the main channel, which is not acceptable.
...
> The binstar client doesn't appear to allow two files with the same name on different channels.

Holy crap.  Talk about "surprising, unexpected behavior".  Can we report this as a bug?

I had no idea---really sorry about this.  The decision to push to a `jenkins` channel was precisely to *avoid* overwriting the packages you and @kyleabeauchamp built.

> The simplest solution might just be to start pushing the builds to the main channel. What are the blockers to doing that?

We can definitely do this, but there may still be a few small loose ends we were tying up.

**Comment by jchodera** [comment]
> The simplest solution might just be to start pushing the builds to the main channel.

Let me know if I should give this a try.

**Comment by jchodera** [comment]
In the meantime, until we've found a solution, I've disabled all the jenkins `conda` builds so we won't overwrite files.

Again, really sorry about this. Had absolutely no idea this was even possible.

**Comment by jchodera** [comment]
Here is the list of things jenkins built successfully on `osx`:
```
+ binstar --quiet -t **** upload --force -u omnia --channel jenkins miniconda/conda-bld/osx-64/ambermini-14.0.1-0.tar.bz2 miniconda/conda-bld/osx-64/cvxopt-1.1.7-py27_0.tar.bz2 miniconda/conda-bld/osx-64/cvxopt-1.1.7-py33_0.tar.bz2 miniconda/conda-bld/osx-64/cvxopt-1.1.7-py34_0.tar.bz2 miniconda/conda-bld/osx-64/cvxpy-0.2.15-np18py27_0.tar.bz2 miniconda/conda-bld/osx-64/docopt-0.6.2-py27_0.tar.bz2 miniconda/conda-bld/osx-64/docopt-0.6.2-py33_0.tar.bz2 miniconda/conda-bld/osx-64/docopt-0.6.2-py34_0.tar.bz2 miniconda/conda-bld/osx-64/ecos-1.0.5-py27_0.tar.bz2 miniconda/conda-bld/osx-64/ecos-1.0.5-py33_0.tar.bz2 miniconda/conda-bld/osx-64/ecos-1.0.5-py34_0.tar.bz2 miniconda/conda-bld/osx-64/fastcluster-1.1.13-np18py27_0.tar.bz2 miniconda/conda-bld/osx-64/fastcluster-1.1.13-np18py33_0.tar.bz2 miniconda/conda-bld/osx-64/fastcluster-1.1.13-np18py34_0.tar.bz2 miniconda/conda-bld/osx-64/fastcluster-1.1.13-np19py27_0.tar.bz2 miniconda/conda-bld/osx-64/fastcluster-1.1.13-np19py33_0.tar.bz2 miniconda/conda-bld/osx-64/fastcluster-1.1.13-np19py34_0.tar.bz2 miniconda/conda-bld/osx-64/fftw3f-3.3.3-1.tar.bz2 miniconda/conda-bld/osx-64/jpype1-0.5.7-np18_0.tar.bz2 miniconda/conda-bld/osx-64/jpype1-0.5.7-np19_0.tar.bz2 miniconda/conda-bld/osx-64/latexcodec-1.0.1-py27_0.tar.bz2 miniconda/conda-bld/osx-64/latexcodec-1.0.1-py33_0.tar.bz2 miniconda/conda-bld/osx-64/latexcodec-1.0.1-py34_0.tar.bz2 miniconda/conda-bld/osx-64/mdtraj-1.2.0-np18py27_0.tar.bz2 miniconda/conda-bld/osx-64/mdtraj-1.2.0-np18py33_0.tar.bz2 miniconda/conda-bld/osx-64/mdtraj-1.2.0-np18py34_0.tar.bz2 miniconda/conda-bld/osx-64/mdtraj-1.2.0-np19py27_0.tar.bz2 miniconda/conda-bld/osx-64/mdtraj-1.2.0-np19py33_0.tar.bz2 miniconda/conda-bld/osx-64/mdtraj-1.2.0-np19py34_0.tar.bz2 miniconda/conda-bld/osx-64/munkres-1.0.7-py27_0.tar.bz2 miniconda/conda-bld/osx-64/munkres-1.0.7-py33_0.tar.bz2 miniconda/conda-bld/osx-64/munkres-1.0.7-py34_0.tar.bz2 miniconda/conda-bld/osx-64/numdifftools-0.6.0-py27_0.tar.bz2 miniconda/conda-bld/osx-64/numdifftools-0.6.0-py33_0.tar.bz2 miniconda/conda-bld/osx-64/numdifftools-0.6.0-py34_0.tar.bz2 miniconda/conda-bld/osx-64/omnia-0.1-0.tar.bz2 miniconda/conda-bld/osx-64/openmm-6.2-py27_0.tar.bz2 miniconda/conda-bld/osx-64/openmmtools-0.0.0-py27_0.tar.bz2 miniconda/conda-bld/osx-64/oset-0.1.3-py27_0.tar.bz2 miniconda/conda-bld/osx-64/oset-0.1.3-py33_0.tar.bz2 miniconda/conda-bld/osx-64/oset-0.1.3-py34_0.tar.bz2 miniconda/conda-bld/osx-64/packmol-1.0.0-0.tar.bz2 miniconda/conda-bld/osx-64/pdbfixer-1.1-np18py27_0.tar.bz2 miniconda/conda-bld/osx-64/pdbfixer-1.1-np19py27_0.tar.bz2 miniconda/conda-bld/osx-64/pybtex-0.18-py27_0.tar.bz2 miniconda/conda-bld/osx-64/pybtex-0.18-py33_0.tar.bz2 miniconda/conda-bld/osx-64/pybtex-0.18-py34_0.tar.bz2 miniconda/conda-bld/osx-64/pybtex-docutils-0.2.0-py27_0.tar.bz2 miniconda/conda-bld/osx-64/pybtex-docutils-0.2.0-py33_0.tar.bz2 miniconda/conda-bld/osx-64/pybtex-docutils-0.2.0-py34_0.tar.bz2 miniconda/conda-bld/osx-64/pyemma-1.0a3-py27_0.tar.bz2 miniconda/conda-bld/osx-64/pymbar-2.1beta0-np18py27_0.tar.bz2 miniconda/conda-bld/osx-64/pymbar-2.1beta0-np19py27_0.tar.bz2 miniconda/conda-bld/osx-64/pystallone-1.0.0b6-py27_0.tar.bz2 miniconda/conda-bld/osx-64/scripttest-1.3-py27_0.tar.bz2 miniconda/conda-bld/osx-64/scripttest-1.3-py33_0.tar.bz2 miniconda/conda-bld/osx-64/scripttest-1.3-py34_0.tar.bz2 miniconda/conda-bld/osx-64/scs-1.0.6-np18py27_0.tar.bz2 miniconda/conda-bld/osx-64/scs-1.0.6-np18py33_0.tar.bz2 miniconda/conda-bld/osx-64/scs-1.0.6-np18py34_0.tar.bz2 miniconda/conda-bld/osx-64/scs-1.0.6-np19py27_0.tar.bz2 miniconda/conda-bld/osx-64/scs-1.0.6-np19py33_0.tar.bz2 miniconda/conda-bld/osx-64/scs-1.0.6-np19py34_0.tar.bz2 miniconda/conda-bld/osx-64/sphinxcontrib-bibtex-0.3.1-py27_0.tar.bz2 miniconda/conda-bld/osx-64/sphinxcontrib-bibtex-0.3.1-py33_0.tar.bz2 miniconda/conda-bld/osx-64/sphinxcontrib-bibtex-0.3.1-py34_0.tar.bz2 miniconda/conda-bld/osx-64/yank-1.0.0-np18_0.tar.bz2 miniconda/conda-bld/osx-64/yank-1.0.0-np19_0.tar.bz2
```
The `linux` version still seems to be having some trouble:
https://jenkins.choderalab.org/job/conda-openmm-release-linux-vagrant/11/
```
BUILD START: openmm-6.2-py27_0
Fetching package metadata: ....
Solving package specifications: ..
Error: Unsatisfiable package specifications.
Generating hint: 
 |                                                                       |   0%
[1/1                 ] |#################################################| 100%
[1/1                 ] |                                                 |   0%
[1/5                 ] |#########                                        |  20%
[2/5                 ] |###################                              |  40%
[3/5                 ] |#############################                    |  60%
[4/5                 ] |#######################################          |  80%
[5/5                 ] |#################################################| 100%
[      COMPLETE      ] |#################################################| 100%

Hint: the following combinations of packages create a conflict with the
remaining packages:
  - sphinxcontrib-bibtex
  - python 2.7*
Build step 'Execute shell' marked build as failure
Warning: you have no plugins providing access control for builds, so falling back to legacy behavior of permitting any downstream builds to be triggered
Finished: FAILURE
```

**Comment by jchodera** [comment]
Whoops, was looking at the wrong jenkins project.  The linux one is working, but I'm not sure it's building a few packages (`msmbuilder` seems to be missing):
https://jenkins.choderalab.org/job/conda-omnia-release-linux-vagrant/13/consoleFull
```
[warning] Distribution linux-64/cvxpy-0.2.15-np18py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/ecos-1.0.5-py33_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/ecos-1.0.5-py34_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/ambermini-14.0.1-0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/docopt-0.6.2-py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/docopt-0.6.2-py33_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/docopt-0.6.2-py34_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/ecos-1.0.5-py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/fastcluster-1.1.13-np18py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/fastcluster-1.1.13-np18py33_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/fastcluster-1.1.13-np18py34_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/fastcluster-1.1.13-np19py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/fastcluster-1.1.13-np19py33_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/fastcluster-1.1.13-np19py34_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/fftw3f-3.3.3-1.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/jpype1-0.5.7-np18_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/jpype1-0.5.7-np19_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/latexcodec-1.0.1-py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/latexcodec-1.0.1-py33_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/latexcodec-1.0.1-py34_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/mdtraj-1.2.0-np18py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/mdtraj-1.2.0-np18py33_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/mdtraj-1.2.0-np18py34_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/mdtraj-1.2.0-np19py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/mdtraj-1.2.0-np19py33_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/mdtraj-1.2.0-np19py34_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/munkres-1.0.7-py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/munkres-1.0.7-py33_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/munkres-1.0.7-py34_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/numdifftools-0.6.0-py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/numdifftools-0.6.0-py33_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/numdifftools-0.6.0-py34_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/omnia-0.1-0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/openmm-6.2-py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/openmmtools-0.0.0-py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/oset-0.1.3-py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/oset-0.1.3-py33_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/oset-0.1.3-py34_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/packmol-1.0.0-0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/pdbfixer-1.1-np18py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/pdbfixer-1.1-np19py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/pybtex-0.18-py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/pybtex-0.18-py33_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/pybtex-0.18-py34_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/pybtex-docutils-0.2.0-py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/pybtex-docutils-0.2.0-py33_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/pybtex-docutils-0.2.0-py34_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/pyemma-1.0a3-py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/pymbar-2.1beta0-np18py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/pymbar-2.1beta0-np19py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/pystallone-1.0.0b6-py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/scripttest-1.3-py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/scripttest-1.3-py33_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/scripttest-1.3-py34_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/scs-1.0.6-np18py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/scs-1.0.6-np18py33_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/scs-1.0.6-np18py34_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/scs-1.0.6-np19py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/scs-1.0.6-np19py33_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/scs-1.0.6-np19py34_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/sphinxcontrib-bibtex-0.3.1-py27_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/sphinxcontrib-bibtex-0.3.1-py33_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/sphinxcontrib-bibtex-0.3.1-py34_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/yank-0.0.0-np18_0.tar.bz2 already exists ... removing 
[warning] Distribution linux-64/yank-0.0.0-np19_0.tar.bz2 already exists ... removing 
```

**Comment by kyleabeauchamp** [comment]
OK, so one workaround is to increment the `build-number` in the conda recipe:

```
build:
  script: $PYTHON setup.py install
  # The build number should be incremented for new builds of the same version
  number: 2
```

This seems to work with a test package: https://binstar.org/kyleabeauchamp/testpackage/files

Ideally, we will have a way to programatically control increment the build number without having to edit the files, however.

**Comment by jchodera** [comment]
There's a "build string" too. I wonder if we could manipulate that as well.

Unfortunately, I don't think either can be can be specified on the conda
build command line. We'd have to check out the recipes and run them through
sed or postprocess with a script, which would be facilitated by having a
build number listed in every recipe.

Jenkins can provide sequential build numbers, but is that enough to keep it
from clobbering manual builds? Unless we pick totally different number
spaces for manual and automated builds?

**Comment by kyleabeauchamp** [comment]
Asked the binstar folks here: https://github.com/Binstar/binstar/issues/109

**Comment by rmcgibbo** [comment]
I'm +1 on pushing all of the binaries from the jenkins servers to `main`. That way, we can focus in on the remaining holdouts that didn't build, and continue to fix their recipes separately. So far I've seen nothing that indicates that any of the built binaries are "incorrect", so I don't see a reason for caution in pushing them to the main channel. The biggest problems so far are things failing to build because of dependency resolution issues, not things being built wrongly.

**Comment by rmcgibbo** [comment]
@jchodera, is it possible to view the script that jenkins is running on say, "https://jenkins.choderalab.org/job/conda-omnia-release-osx-2/"? Is there a config file that is the analog of `.travis.yaml`. The way these things are wired together is still a mystery to me.

**Comment by jchodera** [comment]
> I'm +1 on pushing all of the binaries from the jenkins servers to main. That way, we can focus in on the remaining holdouts that didn't build, and continue to fix their recipes separately.

Sounds good.  I'll re-activate the builds and direct them toward `main`.

> @jchodera, is it possible to view the script that jenkins is running on say, "https://jenkins.choderalab.org/job/conda-omnia-release-osx-2/"? Is there a config file that is the analog of .travis.yaml. The way these things are wired together is still a mystery to me.

The script that gets executed *is* present in the Jenkins output, prefaced with `+` characters.  But I agree that this is not easy to read.

I've just installed the `read-only-configurations` plugin that will allow you to view job configuration info, but I'm still trying to get this to work.

Some time back, I installed a plugin that allows Jenkins to execute `.travis.yaml` files:
https://wiki.jenkins-ci.org/display/JENKINS/Travis+YML+Plugin

It would be great if we could eventually convert everything over to that.  However, we specifically decided not to use this initially because it requires repo-level changes for each repo we work with, and projects like OpenMM actively discourage lots of repo-level changes to muck about with automated building.  However, for `omnia-md/conda-recipes`, this might totally be the way to go. I'm happy to set up a test job if we want to give this a try.

**Comment by rmcgibbo** [comment]
Cool. Mostly I just want to see the scripts, to understand how things work.
Being able to change the config by filing a pull request (e.g. having the
jenkins config stored in the repo in a yaml file or whatever) would be
nice, but it's not essential. I definitely understand why, especially when
you were setting things up, you opted not to go that route.

-Robert

**Comment by jchodera** [comment]
OK, you should have read permission on the job configuration pages:
https://jenkins.choderalab.org/job/conda-omnia-release-linux-vagrant/configure
https://jenkins.choderalab.org/job/conda-omnia-release-osx-2/configure

**Comment by kyleabeauchamp** [comment]
So the Binstar folks pointed out that one can use environment variables inside conda recipes, and one could also have those environment variables set by Jenkins.  That doesn't directly help our current situation, but it could give us a way to get slightly more "revision control" on our binstar-uploaded packages.

**Comment by rmcgibbo** [comment]
I think we should be okay.

After we finish this first release, I'd like to change the jenkins config script such that when a new commit is pushed to this repo, we run `conda build <RECIPE> --output`, and check if the file is already on binstar, and only build the "new" packages. This way we'll basically be rebuilding things incrementally, as new versions are updated, instead of the single shotgun approach where everything needs to be built and pushed in a single job.

**Comment by jchodera** [comment]
> So the Binstar folks pointed out that one can use environment variables inside conda recipes

They suggest we change the recipe to say:
```
build:
  number: {{ environ.get('MY_BUILD_NUMBER', 0) }}
```
I like that solution as a way to ensure we increment the build number with every `jenkins` build of omnia.  We would need to update *all* of the recipes in this repo, however.  Does anybody object to doing this?

@kyleabeauchamp does point out that this doesn't really help our current situation, though. Perhaps we should clarify the issue?

**Comment by jchodera** [comment]
> After we finish this first release, I'd like to change the jenkins config script such that when a new commit is pushed to this repo, we run `conda build <RECIPE> --output`, and check if the file is already on binstar, and only build the "new" packages. This way we'll basically be rebuilding things incrementally, as new versions are updated, instead of the single shotgun approach where everything needs to be built and pushed in a single job.

This sounds reasonable.  

We were not doing this currently because we had found that some conda packages were built incorrectly and needed to be replaced, which would only happen if we (1) did not specify the `omnia` channel during the build (forcing packages to be rebuilt, rather than downloaded), and (2) used the `--force` flag during the `binstar upload` to ensure that we replaced the older packages.

At any point now, I think, we can rework this behavior to instead only upload packages that do not exist on the repo.

Is there a simple way to predict the filename we should check the repo so we don't have to run `conda build` first?

**Comment by rmcgibbo** [comment]
I think this is a mistake. IMO, the correct use case for the build number is when the built binary itself changes. For example, say you released a binary of yank with the problem from #78. You want to release a new binary with the fix. The fix does not involve changing the upstream source of yank, so incrementing the yank version number is not the right thing to do. Instead, you increment the build number.

In our case, most of the time, for most of the packages (whose builds are working), the jenkins machine has simply been rebuilding the same exact binaries over and over. It would be misleading for these to be labeled with increasing build numbers.

**Comment by rmcgibbo** [comment]
> Is there a simple way to predict the filename we should check the repo so we don't have to run conda build first?

Yes, `conda build --output` does this. It doesn't actually build the packages. It just outputs the conda package filename which would have been created and exits. See e.g. https://github.com/omnia-md/conda-recipes/blob/master/conda-build-all#L20-L23
